### DISCUSSION TITLE
Observation on the Clipboard class

### DISCUSSION THREAD

**Discussion by manishpatel0404** [airsdk/Adobe-Runtime-Support#1979] [Ideas]
Recently I came across the need to copy some text from MS Word and paste its HTML (including formatting) into a `TextField`. I used `Clipboard.generalClipboard.getData(ClipboardFormats.HTML_FORMAT)` and received the string containing the HTML markup. However, it was missing all style definitions. Without this information, I could not proceed further. I understand that AS3 `StyleSheet` class supports only a limited number of styles as per the [reference](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/text/StyleSheet.html), which I am okay with. However, if I had access to the style definitions, I could have parsed the Clipboard string and used at least the supported styles, which contain the `font` and `margin` info.

I decided to investigate this further, as this was my first experience with trying to access the Windows Clipboard using code. I learnt that `Clipboard.generalClipboard.getData(ClipboardFormats.HTML_FORMAT)` only provides the markup contained in `<body>` tag.  On digging deeper into the Windows Clipboard structure, I realised that what is being captured is all the text in the `fragment`, i.e. all text between `StartFragment` and `EndFragment`. The style definitions exist in the `<head>` tag above the `<body>` tag. To get the entire HTML text, would require providing all text between `StartHTML` and `EndHTML`. [Info Source](https://docs.microsoft.com/en-us/windows/win32/dataxchg/html-clipboard-format).

If my understanding is correct, is it possible to provide the complete HTML instead of the fragment? It would be useful when applications provide detailed information in the clipboard, such as MS Office apps.

I could write an ANE to achieve this, but I think it would serve more developers if Harman could update this feature since it is already half-way there. Maybe, @ajwfrost  can look into this.

**Comment by ajwfrost** [comment]
Hi

I've just looked at the full data that we get from e.g. Microsoft Word when we open the clipboard with HTML format .. there's a large amount of header information there!

So to get to what you're after, I think there are a couple of options. Simple for us would be to define a new `ClipboardFormat` such as `FULL_HTML_FORMAT` which just returns the full string (so in my test case, this would be a string of 40,000 characters when we're copying 5 characters - the fragment part being 172 characters but as you say, the paragraph 'class' styles are unobtainable currently..). Probably we'd want to actually just return the `<html> ... </html>` part rather than the header - although those values for the offsets would be quite useful...

A more useful (?) mechanism might be to have something like `ClipboardFormat.HTML_CSS` which would just return the contents of the `<style>` tag within the raw HTML. But this would mean you have to call the `Clipboard.getData()` method twice: once for the styles (that could be set into a `StyleSheet` object) and once for the fragment html. I'm not 100% sure whether setting the `htmlText` would then work with the formatting although from what I can see here it should i.e. the `<p class="xyz">` part of the fragment should pick up the style if it's been set via the `TextField.styleSheet` property..

Quick note that this is Windows-specific.. on macOS we just return the full text from the `NSPasteboardTypeHTML` format; on Linux we also return the raw text where the clipboard is provided with `text/html` content.. so in other words, we don't have this issue with other platforms!

Probably the most simple option would be for us to allow the application to decide whether or not it wants the clipboard to give the full html or just the fragment, either via a flag in the application descriptor (which then would be a window-specific flag that changes the behaviour of the existing code, so no changes needed to the AS3 APIs) - or via a new clipboard format that only has relevance on Windows.  But do you have any preferences on this, any thoughts whether the header information is also useful prior to the HTML itself, etc?

thanks

**Reply by manishpatel0404** [reply]
Thanks, @ajwfrost for looking into this.

> A more useful (?) mechanism might be to have something like `ClipboardFormat.HTML_CSS` which would just return the contents of the `<style>` tag within the raw HTML. But this would mean you have to call the `Clipboard.getData()` method twice: once for the styles (that could be set into a `StyleSheet` object) and once for the fragment html. I'm not 100% sure whether setting the `htmlText` would then work with the formatting...

I feel, calling `Clipboard.getData()` method twice may be acceptable, because on the whole setting the style into a `StyleSheet` object would save developer effort of parsing the `<style>` tag. However, even I am not 100% sure about this, because `TextField` may not support all CSS properties within the style.


> Quick note that this is Windows-specific..
> Probably the most simple option would be for us to allow the application to decide whether or not it wants the clipboard to give the full html or just the fragment, either via a flag in the application descriptor (which then would be a window-specific flag that changes the behaviour of the existing code, so no changes needed to the AS3 APIs)

I'm not sure how this might work, but thinking aloud,:

1. Add a property, as you suggest, to `ClipboardFormat` - `ClipboardFormat.HTML_CSS` that would just return the contents of the `<style>` tag within the raw HTML.
2. Add another property, something like `ClipboardFormat.HTML_FRAGMENT` to return the contents of the fragment only.
3. The current `ClipboardFormat.HTML_FORMAT` would return what it currently does (for backward-compatibility), i.e. the full html on non-Windows platforms and fragment on Windows platforms.

This, then, would be usable on all platforms, as well. But, I am not sure how much complexity it would add at your end.

Thanks.

**Reply by ajwfrost** [reply]
Just checking on your suggestions here - the above 1-2-3 list still doesn't include any mechanism for you to get the full `<html>...</html>` contents on Windows? You could get the styles, and the fragment, but may still miss information (e.g. `<body lang=EN-GB style='tab-interval:36.0pt;word-wrap:break-word'>` in my example, or metadata fields such as `<meta name=Generator content="Microsoft Word 15">`).

In terms of getting the CSS information separately, there may be a risk here if you're pulling information from the clipboard twice, as there may be some change in the data between the two calls.. so although the CSS retrieval may help, it may be better to provide some atomic way of pulling out the full HTML content. You can always use the built-in XML parsing to pull out the `<style>` tag contents and then just send this in to the `StyleSheet` parser.

And I'm not 100% sure why you're proposing an `HTML_FRAGMENT` clipboard option .. the only changes that we'd need to make for that would be for the other platforms, to parse the HTML text and pull out the body contents?

thanks

**Reply by manishpatel0404** [reply]
You are absolutely correct on all points. My apologies. I got my head into a muddle on this one.

Your suggestion of adding a Windows-specific flag in the application descriptor seems to be the most appropriate solution.

Thanks.
